src/Map/WeatherTileResourceProvider_P.cpp

```diff
@@ -376,42 +396,221 @@ void OsmAnd::WeatherTileResourceProvider_P::downloadGeoTilesAsync(
 
 std::shared_ptr<OsmAnd::TileSqliteDatabase> OsmAnd::WeatherTileResourceProvider_P::getGeoTilesDatabase()
 {
-    QReadLocker scopedLocker(&_geoDbLock);
+    {
+        QReadLocker geoScopedLocker(&_geoDbLock);
+
+        if (_geoTilesDb)
+            return _geoTilesDb;
+    }
+    {
+        QWriteLocker geoScopedLocker(&_geoDbLock);
 
-    return _geoTilesDb;
+        _geoTilesDb = createGeoTilesDatabase();
+        return _geoTilesDb;
+    }
 }
 
 std::shared_ptr<OsmAnd::TileSqliteDatabase> OsmAnd::WeatherTileResourceProvider_P::getRasterTilesDatabase(BandIndex band)
 {
+    QString key = dateTime.toString(QStringLiteral("yyyyMMdd_hh00"))
+            + QStringLiteral("_")
+            + QString::number(band);
     {
-        QReadLocker scopedLocker(&_rasterDbLock);
+        QReadLocker rasterScopedLocker(&_rasterDbLock);
 
-        const auto citRasterDb = _rasterTilesDbMap.constFind(band);
+        const auto citRasterDb = _rasterTilesDbMap.constFind(key);
         if (citRasterDb != _rasterTilesDbMap.cend())
             return *citRasterDb;
     }
     {
-        QWriteLocker scopedLocker(&_rasterDbLock);
+        QWriteLocker rasterScopedLocker(&_rasterDbLock);
 
         auto db = createRasterTilesDatabase(band);
         if (db)
-            _rasterTilesDbMap.insert(band, db);
-        
+            _rasterTilesDbMap.insert(key, db);
+
         return db;
     }
 }
 
+bool OsmAnd::WeatherTileResourceProvider_P::isEmpty()
+{
+    auto geoDb = getGeoTilesDatabase();
```

Comment:
Not clear why do we need getGeoTilesDatabase instead of _geoTilesDb like before.